lib/pbio/drv/adc/adc_ev3.c

```diff
@@ -48,127 +102,94 @@ pbio_error_t pbdrv_adc_get_ch(uint8_t ch, uint16_t *value) {
+            timer.start += timer.duration;
+        }
 
-    // Reset.
-    SPIReset(SOC_SPI_0_REGS);
-    SPIOutOfReset(SOC_SPI_0_REGS);
+        // Do a sample of all channels
+        pbdrv_block_device_ev3_spi_begin_for_adc(
+            channel_cmd,
+            channel_data,
+            PBDRV_CONFIG_ADC_EV3_ADC_NUM_CHANNELS + PBDRV_CONFIG_ADC_EV3_NUM_DELAY_SAMPLES);
+        PBIO_OS_AWAIT_WHILE(state, pbdrv_block_device_ev3_is_busy());
 
-    // Mode.
-    uint32_t spipc0 = SPI_SPIPC0_SOMIFUN | SPI_SPIPC0_SIMOFUN | SPI_SPIPC0_CLKFUN | SPI_SPIPC0_SCS0FUN3;
-    SPIModeConfigure(SOC_SPI_0_REGS, SPI_MASTER_MODE);
-    SPIPinControl(SOC_SPI_0_REGS, 0, 0, (unsigned int *)&spipc0);
+        for (uint32_t i = 0; i < pbdrv_adc_callback_count; i++) {
+            pbdrv_adc_callbacks[i]();
+        }
+    }
 
-    // Config.
-    SPIClkConfigure(SOC_SPI_0_REGS, SOC_SYSCLK_2_FREQ, 2000000, SPI_DATA_FORMAT0);
-    SPIConfigClkFormat(SOC_SPI_0_REGS, SPI_CLK_OUTOFPHASE | 0x00000010, SPI_DATA_FORMAT0);
-    SPIDelayConfigure(SOC_SPI_0_REGS, 10, 10, 10, 10);
-    SPIIntLevelSet(SOC_SPI_0_REGS, SPI_RECV_INTLVL | SPI_TRANSMIT_INTLVL);
-    SPIDefaultCSSet(SOC_SPI_0_REGS, 8);
+    PBIO_OS_ASYNC_END(PBIO_SUCCESS);
+}
 
-    // Enable and loop around all channels.
-    SPIEnable(SOC_SPI_0_REGS);
+void pbdrv_adc_init(void) {
+    // Immediately go into async mode so that we can wait for the SPI flash driver.
+    // We *don't* want to block the initial init phase, or else things will deadlock.
 
-    process_start(&pbdrv_adc_process);
+    pbio_os_process_start(&pbdrv_adc_ev3_process, pbdrv_adc_ev3_process_thread, NULL);
 }
 
 void pbdrv_adc_update_soon(void) {
-    process_poll(&pbdrv_adc_process);
+    adc_soon = 1;
+    pbio_os_request_poll();
 }
 
-PROCESS_THREAD(pbdrv_adc_process, ev, data) {
-    PROCESS_EXITHANDLER(pbdrv_adc_exit());
-
-    static struct etimer etimer;
-
-    PROCESS_BEGIN();
-
-    etimer_set(&etimer, 10);
-    for (;;) {
-        PROCESS_WAIT_EVENT_UNTIL((ev == PROCESS_EVENT_TIMER && etimer_expired(&etimer)) || ev == PROCESS_EVENT_POLL);
-
-        channel_data_index = 0;
-        adc_busy = true;
-        SPIEnable(SOC_SPI_0_REGS);
-        SPIIntEnable(SOC_SPI_0_REGS, SPI_TRANSMIT_INT);
-        PROCESS_WAIT_EVENT_UNTIL(ev == PROCESS_EVENT_POLL && !adc_busy);
-
-        for (uint32_t i = 0; i < pbdrv_adc_callback_count; i++) {
-            pbdrv_adc_callbacks[i]();
-        }
-
-        etimer_reset(&etimer);
-    }
+void pbdrv_adc_ev3_configure_data_format() {
+    SPIClkConfigure(SOC_SPI_0_REGS, SOC_SYSCLK_2_FREQ, SPI_CLK_SPEED_ADC, SPI_DATA_FORMAT1);
+    // NOTE: Cannot be CPOL=1 CPHA=1 like SPI flash
+    // The ADC seems to use the last falling edge to trigger conversions (see Figure 1 in the datasheet).
+    SPIConfigClkFormat(SOC_SPI_0_REGS, SPI_CLK_POL_LOW | SPI_CLK_OUTOFPHASE, SPI_DATA_FORMAT1);
+    SPIShiftMsbFirst(SOC_SPI_0_REGS, SPI_DATA_FORMAT1);
+    SPICharLengthSet(SOC_SPI_0_REGS, 16, SPI_DATA_FORMAT1);
+}
```

Comment:
I would also set other config explicitly here even if it is the same as flash.

For example, `SPIDefaultCSSet()` is an interesting one. Omitting `SPI_CSHOLD` there is important for using the ADC with DMA so that it toggles the CS line between each word.

Reply:
Which other config settings were you thinking of? `SPIDefaultCSSet` doesn't configure the `CSHOLD` bit. That bit is in fact cleared by the DMA commands to `SPIDAT1`